tests/sync/test_browsertype_connect.py

```diff
@@ -0,0 +1,141 @@
+    assert len(browser1.contexts) == 0
+    browser1.new_context()
+    assert len(browser1.contexts) == 1
+
+    browser2 = browser_type.connect(remote_server.ws_endpoint)
+    assert len(browser2.contexts) == 0
+    browser2.new_context()
+    assert len(browser2.contexts) == 1
+    assert len(browser1.contexts) == 1
+
+    browser1.close()
+    page2 = browser2.new_page()
+    # original browser should still work
+    assert page2.evaluate("7 * 6") == 42
+
+    browser2.close()
+
+
+def test_browser_type_connect_disconnected_event_should_be_emitted_when_browser_is_closed_or_server_is_closed(
+    browser_type: BrowserType, launch_server
+):
+    # Launch another server to not affect other tests.
+    remote = launch_server()
+
+    browser1 = browser_type.connect(remote.ws_endpoint)
+    browser2 = browser_type.connect(remote.ws_endpoint)
+
+    disconnected1 = []
+    disconnected2 = []
+    browser1.on("disconnected", lambda: disconnected1.append(True))
+    browser2.on("disconnected", lambda: disconnected2.append(True))
+
+    page2 = browser2.new_page()
+
+    browser1.close()
+    assert len(disconnected1) == 1
+    assert len(disconnected2) == 0
+
+    remote.kill()
+    assert len(disconnected1) == 1
+
+    with pytest.raises(Error):
+        # Tickle connection so that it gets a chance to dispatch disconnect event.
+        page2.title()
+    assert len(disconnected2) == 1
+
+
+def test_browser_type_disconnected_event_should_have_browser_as_argument(
+    browser_type: BrowserType, launch_server
+):
+    remote_server = launch_server()
+    browser = browser_type.connect(remote_server.ws_endpoint)
+    event_payloads = []
+    browser.on("disconnected", lambda b: event_payloads.append(b))
+    browser.close()
+    assert event_payloads[0] == browser
+
+
+def test_browser_type_connect_set_browser_connected_state(
+    browser_type: BrowserType, launch_server
+):
+    remote_server = launch_server()
+    browser = browser_type.connect(remote_server.ws_endpoint)
+    assert browser.is_connected()
+    browser.close()
+    assert browser.is_connected() is False
+
+
+def test_browser_type_connect_should_throw_when_used_after_is_connected_returns_false(
+    browser_type: BrowserType, launch_server
+):
+    remote_server = launch_server()
+    browser = browser_type.connect(remote_server.ws_endpoint)
+    page = browser.new_page()
+
+    remote_server.kill()
+
+    with pytest.raises(Error) as exc_info:
+        page.evaluate("1 + 1")
+    assert "Playwright connection closed" == exc_info.value.message
+    assert browser.is_connected() is False
+
+
+@pytest.mark.skip
+def test_browser_type_connect_should_reject_navigation_when_browser_closes(
+    server: Server, browser_type: BrowserType, launch_server
+):
+    remote_server = launch_server()
+    browser = browser_type.connect(remote_server.ws_endpoint)
+    page = browser.new_page()
+    server.set_route("/one-style.css", lambda r: None)
+    page.on("request", lambda: browser.close())
```

Comment:
@kumaraditya303 maybe you have an idea here. browser.close blocks in our sync layer, that's why I marked this test as skipped. Will also take a look but maybe you are more experienced with greelet.

Reply:
Will take a look after the tests are fixed

Reply:
should be mostly ready, the only issue which remains that test_browser_type_connect_should_reject_navigation_when_browser_closes timeouts sometimes, will try to debug it later. The other failing tests are general flaky tests which we can fix/investigate also after this PR is merged. Would be nice if there would be more green runs.